docroot/modules/custom/mass_entity_usage/src/Controller/MassLocalTaskUsageController.php

```diff
@@ -0,0 +1,169 @@
+
+    return $build;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function prepareRows($usages) {
+    foreach ($usages as $source_type => $ids) {
+      $type_storage = $this->entityTypeManager->getStorage($source_type);
+      foreach ($ids as $source_id => $records) {
+        // We will show a single row per source entity. If the target is not
+        // referenced on its default revision on the default language, we will
+        // just show indicate that in a specific column.
+        $source_entity = $type_storage->load($source_id);
+        if (!$source_entity) {
+          // If for some reason this record is broken, just skip it.
+          continue;
+        }
+        $field_definitions = $this->entityFieldManager->getFieldDefinitions($source_type, $source_entity->bundle());
+        if ($source_entity instanceof RevisionableInterface) {
+          $default_revision_id = $source_entity->getRevisionId();
+          $default_langcode = $source_entity->language()->getId();
+          $used_in_default = FALSE;
+          $default_key = 0;
+          foreach ($records as $key => $record) {
+            if ($record['source_vid'] == $default_revision_id && $record['source_langcode'] == $default_langcode) {
+              $default_key = $key;
+              $used_in_default = TRUE;
+              break;
+            }
+          }
+        }
+        // If the source is a paragraph, get the parent node.
+        if ($source_entity->getEntityTypeId() == 'paragraph') {
+          /** @var \Drupal\paragraphs\ParagraphInterface $source_entity */
+          $source_entity = $this->getParentNode($source_entity);
+        }
+        $link = $this->getSourceEntityLink($source_entity);
+        // If the label is empty it means this usage shouldn't be shown
+        // on the UI, just skip this row. Also, only show Default sources.
+        if (empty($link) || !$used_in_default) {
+          continue;
+        }
+        // Get the moderation state label of the parent node.
+        $state_label = '';
+        if ($source_entity instanceof Node) {
+          $content_moderation_state = ContentModerationState::loadFromModeratedEntity($source_entity);
+          $state_name = $content_moderation_state->get('moderation_state')->value;
+          $workflow = $content_moderation_state->get('workflow')->entity;
+          $state_label = $workflow->get('type_settings')['states'][$state_name]['label'];
+        }
+        // Get a field label.
+        $field_label = isset($field_definitions[$records[$default_key]['field_name']]) ? $field_definitions[$records[$default_key]['field_name']]->getLabel() : $this->t('Unknown');
+        // Set the row values.
+        $rows[] = [
+          $link,
+          $source_entity->id(),
+          $source_entity->type->entity->label(),
+          $field_label,
+          $state_label,
+        ];
+      }
+    }
+
+    return $rows;
+  }
+
+  /**
+   * Recursive function to retrieve the parent node of a paragraph.
```

Comment:
This looks pretty useful in general. I cant think of a much better place for it though.

Reply:
I copied and modified from some the original entity_usage code that was retrieving a link to the node for paragraph sources. I did something similar in docroot/modules/custom/mayflower/src/Helper.php, but wasn't sure if it was worth it to reference the same method.

Reply:
I think you mentioned on a call or somewhere that it's okay to leave this here as is. Let me know if I'm remember wrong and I can look again.